### PR TITLE
Allow dyn in StagesTuple, add Current Testcase API, Untraitify Progress

### PR DESCRIPTION
**Pull request by domenukk**
Implements proper restarters / progress for Mutational Stages
Also adds new `entry` API to AnyMap,
Adds `.current_testcase` and `.current_input_cloned` APIs
Adds `or_insert_with` to AnyMap, a fast way to insert or get
Changes execution counters from usize to `u64` (for 32 bit systems)
and a lot more...

RFC: Rename `ProgressHelpers` to `RestartHelper`

### CODE REVIEW

File: libafl/src/stages/calibrate.rs [hunk 1]
```diff
@@ -327,6 +327,16 @@ where
 
         Ok(())
     }
+
+    fn initialize_progress(&mut self, state: &mut Self::State) -> Result<(), Error> {
```
Comment (domenukk) [inline libafl/src/stages/calibrate.rs:331]: @addisoncrump _technically_ we could just call this from the stage, not sure what's best

File: libafl/src/stages/dump.rs [hunk 2]
```diff
@@ -115,6 +113,18 @@ where
 
         Ok(())
     }
+
+    #[inline]
+    fn restart_progress_should_run(&mut self, _state: &mut Self::State) -> Result<bool, Error> {
```
Comment (domenukk) [inline libafl/src/stages/dump.rs:118]: Don't 100% like the name but it's the best I found

File: libafl/src/stages/colorization.rs [hunk 3]
```diff
@@ -93,6 +99,16 @@ where
 
         Ok(())
     }
+
+    fn restart_progress_should_run(&mut self, state: &mut Self::State) -> Result<bool, Error> {
+        // TODO this stage needs a proper resume
+        RetryRestartHelper::restart_progress_should_run(state, self, 3)
```
Comment (tokatoka) [inline libafl/src/stages/colorization.rs:105]: should colorization stage also have self.restart_helper ?

File: libafl/src/stages/logics.rs [hunk 4]
```diff
@@ -3,36 +3,31 @@
...
-{
-    fn initialize_progress(state: &mut S, _stage: &ST) -> Result<(), Error> {
+impl NestedStageRestartHelper {
+    fn restart_progress_should_run<S, ST>(state: &mut S, _stage: &ST) -> Result<bool, Error>
+    where
+        S: HasNestedStageStatus,
+    {
         state.enter_inner_stage()?;
-        Ok(())
+        Ok(true)
```
Comment (tokatoka) [inline libafl/src/stages/logics.rs:21]: this is not always true? right?
i think it is depending the result of the inner stage

File: libafl/src/stages/mod.rs [hunk 5]
```diff
@@ -336,6 +447,193 @@ where
...
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ExecutionCountRestartHelperMetadata {
+    /// How many executions we had when we started this stage initially (this round)
+    started_at_execs: u64,
+}
+
+/// A tool shed of functions to be used for stages that try to run for `n` iterations.
+///
+/// # Note
+/// This helper assumes resumable mutational stages are not nested.
```
[the remark this answers is not in the document]
Comment (domenukk) [reply libafl/src/stages/mod.rs:573]: Yes exactly. This means overhead for the common case (but worst case we'll have to change it in the future)

**Comment by domenukk** [comment]
Replied to the comments

**Comment by domenukk** [comment]
No, I don't understand how to reply :D

**Comment by domenukk** [comment]
The 'dyn doesn't help us too much here because the stages need to be `'static`

File: libafl/src/stages/logics.rs (hunk 4 above)
Comment (domenukk) [reply libafl/src/stages/logics.rs:21]: Don't think so

**Comment by tokatoka** [comment]
I think it's confusing that Stage implements functions named `restart_progress_should_run` and helpers also implement `restart_progress_should_runs`.
Shouldn't we rename one of them?

File: libafl/src/stages/mod.rs [hunk 6]
```diff
@@ -336,6 +447,193 @@ where
...
+    #[inline]
+    fn clear_restart_progress(&mut self, _state: &mut Self::State) -> Result<(), Error> {
+        Ok(())
+    }
+}
+
+/// Progress which permits a fixed amount of resumes per round of fuzzing. If this amount is ever
+/// exceeded, the input will no longer be executed by this stage.
+#[derive(Clone, Deserialize, Serialize, Debug)]
+pub struct RetryRestartHelper {
```
Comment (tokatoka) [inline libafl/src/stages/mod.rs:466]: should we make a 
```
pub trait RestartHelper
 fn restart_progress_should_run(); 
 fn clear_restart_progress();
}
```
trait?

File: libafl/src/stages/colorization.rs (hunk 3 above)
Comment (tokatoka) [reply libafl/src/stages/colorization.rs:105]: ok. i think the current `RetryRestartHelper` is fine

File: libafl/src/stages/logics.rs (hunk 4 above)
Comment (tokatoka) [reply libafl/src/stages/logics.rs:21]: hm okay i understand.

but i noticed one another problem.
These conditional stage evaluates the closure given by the user. but you could get stuck when evaluating them right? so how about also adding max # of retry to this helper too?

Comment (domenukk) [reply libafl/src/stages/logics.rs:21]: ClosureStage uses the RetryHelper. For all user-provided stages, they'll have to implement their own restart handling

Comment (domenukk) [reply libafl/src/stages/logics.rs:21]: I don't think bailing after `n` retries will always be the best solution (I would even say it's always the worst)

File: libafl/src/stages/mod.rs (hunk 6 above)
Comment (domenukk) [reply libafl/src/stages/mod.rs:466]: It exists, but some helpers need different parameters so the trait doesn't apply

**Pull request merged**